cmd/derper/derper.go

```diff
@@ -212,25 +213,14 @@ func main() {
 		tsweb.AddBrowserHeaders(w)
 		w.Header().Set("Content-Type", "text/html; charset=utf-8")
 		w.WriteHeader(200)
-		io.WriteString(w, `<html><body>
-<h1>DERP</h1>
-<p>
-  This is a <a href="https://tailscale.com/">Tailscale</a> DERP server.
-</p>
-<p>
-  Documentation:
-</p>
-<ul>
-  <li><a href="https://tailscale.com/kb/1232/derp-servers">About DERP</a></li>
-  <li><a href="https://pkg.go.dev/tailscale.com/derp">Protocol & Go docs</a></li>
-  <li><a href="https://github.com/tailscale/tailscale/tree/main/cmd/derper#derp">How to run a DERP server</a></li>
-</ul>
-`)
-		if !*runDERP {
-			io.WriteString(w, `<p>Status: <b>disabled</b></p>`)
-		}
-		if tsweb.AllowDebugAccess(r) {
-			io.WriteString(w, "<p>Debug info at <a href='/debug/'>/debug/</a>.</p>\n")
+		err := homePageTemplate.Execute(w, templateData{
+			ShowAbuseInfo: validProdHostname.MatchString(*hostname),
+			Disabled:      !*runDERP,
+			AllowDebug:    tsweb.AllowDebugAccess(r),
+		})
+		if err != nil {
+			http.Error(w, err.Error(), 500)
```

Comment:
It's almost always too late for this, as the template has probably written. And usually the error is because the client disconnected.

More common is:

    if err != nil {
          if r.Context().Err() == nil {
                log.Printf("homePageTemplate.Execute: %v", err)
          }
          return
      }

Plus the "500" part is already ignored due to the explicit 200 on line 215.

If you want to check that the template+data are valid, you need to write it to a `bytes.Buffer` first and then either send the http.Error(500) or w.Write(buf.Bytes())

Reply:
Good call.